src/twisted/logger/test/_assertionhelpers.py

```diff
@@ -0,0 +1,244 @@
+# -*- test-case-name: twisted.logger.test.test_assertionhelpers -*-
+# Copyright (c) Twisted Matrix Laboratories.
+# See LICENSE for details.
+
+"""
+General assertion helpers for testing log events.
+"""
+
+from contextlib import contextmanager
+
+from twisted.logger import globalLogPublisher
+from twisted.logger._legacy import _ensureCompatWithLegacyObservers
+
+
+
+@contextmanager
+def capturingLogEventsFrom(logger):
+    """
+    Context manager to accumulate log events emitted by a L{Logger}
+    instance by passing list.append as an observer.
+
+    @param logger: L{Logger} object that the event was emitted to.
+    @type logger: L{Logger}
+    """
+    events = []
+    logger.observer = events.append
+    if logger.source is not None:
+        logger.source._log = logger
+
+    yield events
+
+
+
+@contextmanager
+def capturingLogEventsFromGlobalLogPublisher():
+    """
+    Context manager to accumulate log events emitted by a L{Logger}
+    instance that is not directly accessible for testing, such as
+    when L{twisted.logger._loggerFor} is used to create a L{Logger}
+    on the fly. This is to avoid creating L{Logger} instance attributes in
+    the Twisted codebase itself to avoid namespace conflicts for when end
+    users subclass. In this case we don't assert log_logger in the event
+    dict and ignore it and accumulate the events in a list by passing
+    list.append as an observer to L{twisted.logger.globalLogPublisher}.
+    In the event that the L{Logger} is an instance attribute,
+    L{capturingLogEventsFrom} should be used in conjunction with
+    L{assertGlobalLogEvent} and L{assertGlobalLogEvents} for optimum test
+    coverage. Note that this is preferrably only for use in Twisted itself
+    to facilitate default logging.
+    """
+    events = []
+    globalLogPublisher.addObserver(events.append)
+
+    yield events
```

Comment:
This should be:

```python
try:
    yield events
finally:
    globalLogPublisher.removeObserver(events.append)
```

That way if an exception occurs in the context manager's block, the observer will still be removed.